### ISSUE TITLE
[bg] ALPHA TEST: Backgrounded MIME type loading, dir sizes, mount point refresh

### ISSUE THREAD

**Issue by IgnorantGuru** [IgnorantGuru/spacefm#582]
For those who would like to help with alpha testing, there is a temporary 'bg' branch which includes some changes to dir loading and features.  Because this branch makes extensive changes to multi-threading behavior, it may crash or hang and can use some heavy testing in assorted environments.  However, aside from UI instability it should be relatively safe to use.

To install this branch, you can run the [net installer](http://ignorantguru.github.io/spacefm/#installer):

    spacefm-installer --version=bg

This branch includes the following changes:

<ul>
    <li>How SpaceFM loads a directory has changed.  <b>Previously</b>, it loaded the directory like this:<br><br>
    <ul>
        <li>Background Thread 1: Load the file list and file properties, and load the MIME types of all files (which involves opening/reading some files) and associated icons
        <li>UI Thread: Show the file list and let user perform functions.
        <li>Background Thread 2: Load thumbnails
    </ul>
    <br><br>
    The above loading was slow when opening dirs like /usr/bin because the MIME type loading was slow.
    <br><br>
    The <b>new dir loading</b> is performed like this:
    <ul>
        <li>Background Thread 1: Load the file list and file properties.
        <li>UI Thread: Show the file list and let user perform functions.
        <li>Background Thread 1: Load MIME Types and shows the icons.
        <li>UI Thread: When MIME Types are done loading, if sort is set to Type, the list will be resorted.
        <li>Background Thread 1: Calculate dir sizes
        <li>Background Thread 2 (started at same time as 'Calculate dir sizes' in another thread): Load thumbnails
        <li>UI Thread: When dir sizes are done loading, if sort is set to Size, the list will be resorted.
    </ul>
    <br>
    Iow, the file list should be displayed almost immediately and you can begin using it as MIME types and icons are loaded, followed by dir sizes and thumbnails.  (Issue #544)<br><br>
    <li>Except for non-block filesystems listed in Devices|Settings|Change Detection, SpaceFM will calculate the deep size of subdirectories listed in the current directory.  It will not list size for /proc, /mnt, or /sys.  It will follow links in the current directory, but will not follow links within subdirectories.  It will include the size of subdirectories mounted on another device, but will not includes files within the subdirectory which are mounted on another device.  Thus it may differ from the size given by Properties|Info.  It will roughly match the command 'du --apparent-size -hx'.  (Thus if /media/cdrom is mounted, and the current directory is '/', the 'media' subdir will not include the size of /media/cdrom.  Yet if you enter /media, then the size of 'cdrom' will be shown.)  If subdirs cannot be read due to permissions, their size will not reflect contents.  Dir sizes are not updated automatically - you must refresh or reopen the tab to get updated sizes.  (Issue #338)<br><br>
    
    <li>The View|Refresh function has been redesigned.  If multiple tabs are open to the same dir, refreshing one will now refresh all of them.  (Previously, SpaceFM was unable to reload the dir for refresh if multiple tabs were open to the same dir.)<br><br>
    
    <li>When mounting or unmounted a device, any tabs already open to the (canonicalized) mount point will be automatically refreshed.  They can also now be refreshed manually even if multiple tabs are open to the same mount point.  (Issue #568)<br><br>
    
    <li>Currently for testing, some additional output will be seen on stdout.  If SpaceFM crashes or hangs, please choose the Build Debug option in the installer, or see README.
</ul>

You can [see recent changes to bg branch](https://github.com/IgnorantGuru/spacefm/commits/bg), and also [see all code changes between this version and next](https://github.com/IgnorantGuru/spacefm/compare/next...bg).

If you have any questions or issues, please comment below.  Also, feedback is welcome on the following:

Do you think there should be a global option to disable dir sizes?  This option seems very useful and performs well, so thus far I haven't made it an option.  If you think it should be optional, present your case and it will be considered (maybe try it for awhile first).  Same for backgrounded MIME Type loading - I don't see any reason to make it optional.

Thanks for testing!

**Comment by IgnorantGuru** [comment]
A new run-task socket task type <a href="http://ignorantguru.github.io/spacefm/spacefm-manual-next-en.html#run-task-fresh">refresh</a> has been added to refresh tabs or dirs.

**Comment by morfikov** [comment]
I've tested this version, and now the issue #568 is gone completely. The content of a mount point appears in the open tab after mount the partition.

**Comment by IgnorantGuru** [comment]
**Update:** Mostly this branch is stabilized, but there are a few remaining issues.  Mostly these only appear in stress tests where many tabs are being refreshed or chdir very quickly and repetitively, in which case it might print a warning to stdout or in very rare cases crash.

I'd appreciate as many testers as possible.  For general use, this now seems very stable.  You can just note any warnings on stdout or any actions that cause a problem.

**Comment by jmcamp** [comment]
One suggestion:  since there's now a background thread to calculate directory sizes, can #245 be implemented within the same background thread with no extra performance penalty?  (https://github.com/IgnorantGuru/spacefm/pull/245)

**Comment by IgnorantGuru** [comment]
@jmcamp They're not really related.  You're welcome to open a new request on that or add a comment to #17, which I would review before doing any work in that area, so it can be used to collect ideas for the dir tree.

**Comment by IgnorantGuru** [comment]
The above changes in the bg branch will not be included in the upcoming SpaceFM 1.0.4 release because of some remaining stability issues to be addressed.  The bg branch has been merged into an 'alpha' branch, which can be used to continue alpha-testing this version.

I plan to use the 'alpha' branch from time to time for this kind of testing, rather than always using a different test branch name.  However, the alpha branch may not always exist, so any build scripts should fall back to the 'next' branch if 'alpha' is missing.

This bg/alpha branch should be fine to use for most users - the instability will rarely if ever show itself through normal use.  If it does, it may just cause a crash or hang.  If you do encounter one, any details are welcome, particularly any warnings in stdout.

**Comment by paoloschi** [comment]
> Background Thread 1: Load the file list and file properties, and load the MIME types of all files (Which Involves opening / reading some files) and associated icons

here with 1.0.4+alpha if I list *.mod files in `/boot/grub/i386-pc` the description still says "Amiga Soundtracker"

expected behavior? or a misconfiguration of my system?

**Comment by IgnorantGuru** [comment]
@paoloschi There haven't been any significant changes to how MIME types are determined, only when they're read in the dir loading process.  *.mod files identify that way on my system too.  This is done via glob in the file /usr/share/mime/audio/x-mod.xml.  If you copy that file to ~/.local/share/mime/packages/audio-x-mod.xml and edit the comment (description), then restart SpaceFM, it should be changed for your current user in all apps.

Or, if you associate an application with that file type, then right-click on the application in SpaceFM's Open context menu, you can use the [MIME Menu](http://ignorantguru.github.io/spacefm/spacefm-manual-en.html#designmode-mime) to copy that file, but you'll still need to edit it manually.

Identifying the file as an Amiga file by default is probably a bug or oversight in [mime-shared-info](http://freedesktop.org/wiki/Software/shared-mime-info/) library (Freedesktop), so that would be the upstream place to file a bug or request for change.

Thanks for testing.

**Comment by Vladimir-csp** [comment]
Will the new threads system help reducing GUI hangs during long i/o wait?

**Comment by IgnorantGuru** [comment]
@Vladimir-csp It can help reduce the time waiting for dir load, since more of it is now backgrounded (the MIME type identification requires opening some files, and this is now done in a background thread).  So when I test with curlftpfs, for example, I notice the dir shows faster, then the MIME types are gradually loaded.  So it seems more responsive.

However, aside from dir loading, stat() and other access calls will still hang.  So clicking on a file in an unresponsive fs will still hang until the fs responds.  In general, SpaceFM is as responsive as the filesystem being worked on.  This change doesn't add asynchronous i/o for that GUI-related access.

Also note that dir size loading obeys the Devices|Settings|Change Detection blacklist.

**Comment by IgnorantGuru** [comment]
**Update:**  Problems with thumbnail loading (where a mouseover or scroll was required for the thumbnail to be updated, discussed elsewhere) should now be corrected.

**Changed:**  A change has been made to the Change Detection blacklist in this branch, such that when a filesystem is blacklisted, newly created files will now be detected and read for their properties and MIME type.  Deleted files have always been detected as well.  Now only file content or property changes are ignored.  This will make SpaceFM show files that are created, copied, renamed or deleted.  Feedback is welcome on how this performs - is this desirable?  The cost is that on less responsive filesystems, creating files may cause a hang until the new files are read.

**Comment by IgnorantGuru** [comment]
I have made some improvements to the status bar - now shows total dir content size if no files are selected (this may differ from the dir size you see before navigating into the dir, because it only counts the files in the dir, not the dir itself), show derefed link sizes if single selection, and for multiple selection it shows the actual total (if you copied the selected files, this is how much data you're copying).  This allows Icon users to click on a single dir or link to see the deep size.  Comments welcome.

Also, if SpaceFM can't open a dir, no size is shown now, instead of 4.0K.  (If files or dirs within the dir cannot be accessed, this is ignored with the total showing what can be read.)

This branch is just about ready to go into next so any final testing or routine use is appreciated!

**Comment by Vladimir-csp** [comment]
Works good so far. A little cosmetic issue on the status bar: IMHO, it is better to not show `(0 B)` at all when dir with unknown size is selected.

**Comment by IgnorantGuru** [comment]
@Vladimir-csp 
> better to not show (0 B) at all when dir with unknown size is selected.

The size is not unknown per se.  Rather it should show the accurate amount visible, just like if you run 'du -h /root'.  But 0B is incorrect, should be something like 4.0K.  I don't think it looks good to show nothing.

I have made a change where it now shows both the apparent size and the deep size when links are selected.  And if a dir has zero size (no size shown in the list, indicating the size hasn't loaded yet or failed to open), then the lstat size is used (eg 4.0 K).  Dirs like /sys (which aren't real) may still show 0 B, just like du does.

Long story short, the first amount shown is the amount that would be copied if you copied the selected files (roughly the same as du --apparent-size -hx DIR).  The second number follow links (but not deep links).  I think it's handy to be able to add up the defered dir links in some cases, so I think both numbers are useful, and seeing the arrow reminds you that some of what is selected is links.

Thanks for testing.

**Comment by IgnorantGuru** [comment]
This alpha branch has been merged into next, due for 1.0.5 release.  Any related problems can be noted below or opened in a new issue.

**Comment by tome-** [comment]
I think there should be an option to disable dir sizes for example View -> Columns -> Size (files) and second Size (files and dirs). Calculating and sorting dir sizes may take some time on slower devices so you can click something quickly but after a while it is something else.

**Comment by IgnorantGuru** [comment]
If I add an option it would a global option in Preferences, as the dir size calc happens on the directory level which may impact multiple panels and windows.  It would be like the Show Thumbnails option, perhaps putting it in View also so it can associated with keyboard and put on the toolbar.

Have you encountered any real-use difficulty with it, or are you hypothesizing?  I tend to agree that some people may want the option, but I expected it to be more demanding than it is - it seems to work fluidly in the circumstances I've tested it.

**Comment by tome-** [comment]
It is real, I use 10 years old laptop and I hear my disk while it is reading size, something like short circuit. Before sorting by size dirs are for a (too long) while sorted by name. I am hypothesizing only solutions, for example global option "Show dirs sizes" or button/key - "Calculte dirs sizes only for current tab/window".

**Comment by IgnorantGuru** [comment]
**Update:**  Due to remaining instability in these feature changes, these changes have been unmerged from the next branch and are once again available via **only the alpha branch**.

While the alpha branch should work fine most of the time, and I encourage use of this branch, it does not quite meet the quality control for release due to very rare crashes or hangs, eg #602.

While most of the issues causing instability have been resolved, there are still a few hard-to-find races that seem to be occurring between threads.  Feedback on your results welcome, and I'll be continuing to use and bugfix this branch, getting it to release asap.

**Comment by IgnorantGuru** [comment]
**Update:  A Show Folder Sizes option has been added** to Preferences next to Show Thumbnails.  This global option also appears in View|Style|Folder Sizes (where you can set a shortcut key), and you can add it as a new builtin toolbar item.  There is also a new matching show_dirsize socket property.

When toggling this option, all open dirs will be refreshed.  With the option disabled, no deep dir sizes should be calculated or shown.

**Comment by IgnorantGuru** [comment]
All known crashes in alpha should now be corrected as of f91c4b4  If anyone does still see any crash or GTK warnings from alpha, please report it again in this issue or elsewhere.

To obtain a complete backtrace, please see the BUILD DEBUG instructions in the [updated README](https://raw.githubusercontent.com/IgnorantGuru/spacefm/1805a151af27a0d4c1f03247e9751b0813564d66/README).

**Comment by toweih** [comment]
Is there any chance the dir-deep-size-feature will be back some day?
